src/core/textrenderer/qgstextrenderer.cpp

```diff
@@ -150,6 +152,215 @@ void QgsTextRenderer::drawText( QPointF point, double rotation, Qgis::TextHorizo
 }
 
+void QgsTextRenderer::drawTextOnLine( const QPolygonF &line, const QString &text, QgsRenderContext &context, const QgsTextFormat &format, double offsetAlongLine, double offsetFromLine )
+{
+  QgsTextFormat tmpFormat = format;
+  if ( format.dataDefinedProperties().hasActiveProperties() ) // note, we use format instead of tmpFormat here, it's const and potentially avoids a detach
+    tmpFormat.updateDataDefinedProperties( context );
+  tmpFormat = updateShadowPosition( tmpFormat );
+
+  // todo handle newlines??
+  QgsTextDocument document = format.allowHtmlFormatting() ? QgsTextDocument::fromHtml( { text  } ) : QgsTextDocument::fromPlainText( { text  } );
+  document.applyCapitalization( format.capitalization() );
+
+  drawDocumentOnLine( line, tmpFormat, document, context, offsetAlongLine, offsetFromLine );
+}
+
+void QgsTextRenderer::drawDocumentOnLine( const QPolygonF &line, const QgsTextFormat &format, const QgsTextDocument &document, QgsRenderContext &context, double offsetAlongLine, double offsetFromLine )
+{
+  QPolygonF labelBaselineCurve = line;
+  if ( !qgsDoubleNear( offsetFromLine, 0 ) )
+  {
+    QgsGeometry curve = QgsGeometry::fromQPolygonF( line );
+    labelBaselineCurve = curve.offsetCurve( offsetFromLine, 4, Qgis::JoinStyle::Round, 2 ).asQPolygonF();
+  }
+
+  const double fontScale = calculateScaleFactorForFormat( context, format );
+
+  const QFont baseFont = format.scaledFont( context, fontScale );
+  const double letterSpacing = baseFont.letterSpacing() / fontScale;
+  const double wordSpacing = baseFont.wordSpacing() / fontScale;
+
+  QStringList graphemes;
+  QVector< QgsTextCharacterFormat > graphemeFormats;
+  QVector< QgsTextDocument > graphemeDocuments;
+  QVector< QgsTextDocumentMetrics > graphemeMetrics;
+
+  for ( const QgsTextBlock &block : std::as_const( document ) )
+  {
+    for ( const QgsTextFragment &fragment : block )
+    {
+      const QStringList fragmentGraphemes = QgsPalLabeling::splitToGraphemes( fragment.text() );
+      for ( const QString &grapheme : fragmentGraphemes )
+      {
+        graphemes.append( grapheme );
+        graphemeFormats.append( fragment.characterFormat() );
+
+        QgsTextDocument document;
+        document.append( QgsTextBlock( QgsTextFragment( grapheme, fragment.characterFormat() ) ) );
+
+        graphemeDocuments.append( document );
+        graphemeMetrics.append( QgsTextDocumentMetrics::calculateMetrics( document, format, context, fontScale ) );
+      }
+    }
+  }
+
+  QVector< double > characterWidths( graphemes.count() );
+  QVector< double > characterHeights( graphemes.count() );
+  QVector< double > characterDescents( graphemes.count() );
+  QFont previousNonSuperSubScriptFont;
+
+  for ( int i = 0; i < graphemes.count(); i++ )
+  {
+    // reconstruct how Qt creates word spacing, then adjust per individual stored character
+    // this will allow PAL to create each candidate width = character width + correct spacing
```

Comment:
Is the PAL reference here correct? I assume this function will not be used within PAL as it handles curves separately?